tox.ini

```diff
@@ -18,13 +18,13 @@ deps =
   ; we have a brain for nose
   ; we use pytest for tests
   nose
-  py35,py36,py37,py38,py39: numpy
-  py35,py36,py37,py38,py39: attr
-  py35,py36,py37,py38,py39: typed_ast>=1.4.0,<1.5
+  py{35,36,37,38,39}: numpy
+  py{35,36,37,38,39}: attrs
+  py{35,36,37,38,39}: typed_ast>=1.4.0,<1.5
   pytest
-  python-dateutil
+  !py{35,36,37,38,39}-six: python-dateutil
   pypy: singledispatch
-  six~=1.12
+  six: six
```

Comment:
@dgilman i'am sorry but i don't understand how do you ensure that in environments suffixed with `-six` the `six` module is really absent...
In fact it seems to be the opposite. But can you explain how do you manage this?

Reply:
I am not sure what to say about the -tox line, that seems to be incorrect and I will fix it, but tox appears to not care. The root of the configuration happens on this line though, and is documented upstream here https://tox.readthedocs.io/en/latest/config.html#factors-and-factor-conditional-settings

Just by declaring `six: six` a new factor called six has been created (the factor name is the part before the colon), and any tox environment that contains that factor will install the dependency.

The setUpClass that raises an exception is only run when you're under tox (`TOX_ENV_NAME` envvar is defined), and six is installed (`HAS_SIX`), and the six factor was not declared in this environment (`tox_env.endswith('-six')`). If you run the unit test case locally you won't have TOX_ENV_NAME declared, so you don't have to worry about this conditional. If you are running under tox you have tox setting up and tearing down the virtualenv, which because of the above `deps=` configuration makes sure six is or is not installed, allowing us to do an extra round of double checking that six did not sneak in either as an explicit dependency in our tox.ini or because one of our dependencies decided to pull it in. That's why python-dateutil is blacklisted for six, it's pulling in six for its own purposes.

Reply:
@dgilman thanks a lot for this explanation and the like toward the doc. Now i understand. :+1: